### ISSUE TITLE
🚩 Add support for more trigger events

### ISSUE THREAD

**Issue by lindseydiloreto** [doublesecretagency/craft-notifier#2]
At the moment, only a current event is supported:

 - On Save Entry

This can (and should) be easily expanded to cover other events:

 - On Delete Entry
 - On Save Asset/User/etc.
 - On Delete Asset/User/etc.

I'm not sure when each particular event will be supported... Feel free to leave a comment if you have a preference! 🙂

**Comment by AndrewMac** [comment]
Whatever you do for entries, I'd like to see for Commerce Products, Digital Products and Users. Thanks! :-)

**Comment by lindseydiloreto** [comment]
Noted! 🙂

**Comment by aaronmclayton** [comment]
Is on save/entry and publish entry the same? One could save to a draft, and technically save. I'm not sure the controller actions on this in Craft.

**Comment by lindseydiloreto** [comment]
That's a great question. _All_ saved entries will trigger this event, so it's really a matter of filtering to get the behavior that you want...
<img width="656" alt="filters@2x" src="https://user-images.githubusercontent.com/5309692/131228817-481d76d5-62d4-42c6-b5da-3140ad15ef8a.png">

 - The "Freshness" filter determines whether the entry is brand new [(via `isNew`)](https://github.com/doublesecretagency/craft-notifier/blob/c9f3ac93fb47859fa04122c5626836d1a9571774/src/models/Trigger.php#L63-L73).
 - The "Drafts" filter determines whether the entry is a [Draft or Published](https://github.com/doublesecretagency/craft-notifier/blob/c9f3ac93fb47859fa04122c5626836d1a9571774/src/models/Trigger.php#L75-L85).

There are a lot more subtleties and nuances to the "On Entry Save" event, including plenty of stuff that I'm still trying to wrap my head around. It's quite possible that these two filters will evolve into something more significant down the road.

**Comment by theurbanu** [comment]
Do you plan on adding the ability to further filter saved new entries, for example, if entry is new AND a specific lightswitch field in enabled?

Another example would be if the status of a entry is changed from pending to enabled, can an email be sent to the author of that entry?

**Comment by lindseydiloreto** [comment]
> "... if entry is new AND a specific lightswitch field is enabled"
> "... if the status of an entry is changed from pending to enabled"

The very short answer is "yes". All of those features are on the roadmap. 🙂

The longer answer... as a workaround, you can take advantage of these two existing features to achieve your goals...

 - [`skipMessage`](https://plugins.doublesecretagency.com/notifier/messages/skip/) - Within the context of a Twig message, you can create simple "yes/no" logic. If the conditions fail any of your tests, you can easily `skipMessage` and move on to the next one.
 - [Automatic variables](https://plugins.doublesecretagency.com/notifier/messages/variables/) - These variables will automatically be available in your Twig templates. You can use them to determine whether or not to `skipMessage`.

Here's some rough pseudo-code...

```twig
{% if entry.myField != 'value that I want' %}
    {% skipMessage %}
{% endif %}
```

---

> "... can an email be sent to the author of that entry?"

Yup, absolutely! When editing the Message in the control panel, select [Custom > Users](https://plugins.doublesecretagency.com/notifier/recipients/custom-users/) and specify the **author** of each Entry.

```twig
{# Encode the Entry author's ID #}
{{ [entry.author.id]|json_encode }}
```

^ Fair warning, we may change the "Custom" behavior before 1.0 is released. Handling this logic in Twig is complicated and clunky, we may move the customization logic into PHP.

**Comment by mach1media** [comment]
Agreed with @AndrewMac - Whatever you do for entries, I'd like to see for Commerce Products, Digital Products and Users. 😃

**Comment by lindseydiloreto** [comment]
Just got another vote for User events and Solspace Calendar events.

Hoping to get to this soon (for anyone following this thread). 🤞

**Comment by lindseydiloreto** [comment]
+1 request for the "on save user" event (per @darylknight).

**Comment by johnj24** [comment]
Lindsey, is an asset upload trigger possible from a front-end user form?

**Comment by lindseydiloreto** [comment]
@johnj24 I believe so. You want a trigger for **when the file is uploaded** (not when the Asset is saved).

Is that correct?

**Comment by johnj24** [comment]
@lindseydiloreto I guess ideally both, but for now I think all I'm looking for is on upload. I have a client looking to be notified when a user uploads a file through a front-end form. The user does not have the ability to save any file changes beyond that, if that's what you mean?

**Comment by jl-digital** [comment]
I'd like the event to only fire on the set post date and time. If I set the post date of an entry in the future, the email will be sent on entry save, not on entry postdate. But the entry is not yet visible on the front end.

**Comment by timprint** [comment]
I am currently using Percipio's "Notifications" plug-in but the Craft 3 version is incompatible with PHP 8.1 and 8.2. I am looking for an alternative.

I am using it to send emails to users when they are added to specific user groups.

For example: "_You're membership has been approved, your discounts will automatically be applied at checkout_" when an admin manually adds them to a specific user group.

Is that a possibility for v.1 ?

**Comment by lindseydiloreto** [comment]
Hi @timprint, great question. Yes, that feature will definitely be possible in v.1. 👍 

I've been doing a lot of work on v.1 lately, there have been some significant changes since the original beta version. Feel free to ping me on Discord if you want to get a sneak peek!

**Comment by lindseydiloreto** [comment]
Excellent news, I believe all of these various trigger events are included in [v3.0.0](https://github.com/doublesecretagency/craft-notifier/releases/tag/3.0.0)! 🚀 👏 

Read the docs for the wide variety of [event types](https://plugins.doublesecretagency.com/notifier/events/types/) now available...

<img width="416" alt="All Event Types" src="https://plugins.doublesecretagency.com/notifier/images/events/event-types.png" />

Hope that helps, let me know if I missed anything! 🙏
